rfd/0139-gogo-must-go.md

```diff
@@ -116,12 +151,128 @@ casts or conversions within go code. We will have to take on this conversion wor
+today.
 
 ##### `gogoproto.stdtime`
 
 We will need to migrate to using `timestamppb` directly instead of relying on gogo's `stdtime` tag.
 
+### Proposed solution
+
+#### Create internal objects
+
+Currently, in `api/types` we define a large number of interfaces that are subsequently used to wrap
+our protobuf created messages in a nicer, more user friendly way, abstracting away much of the protobuf
+bits and pieces.
+
+We should break this relationship and create actual internal objects for our interfaces. From here, we
+should use conversion functions to translate between protobuf messages and our internal objects.
+
+It's imperative that these objects are compatible with the current JSON/YAML representations in Teleport.
+That is, the objects that are in Teleport today should unmarshal properly into our new internal objects.
+
+All objects should remain in `api/types` as they do today. (Open question: should this be the case? Should
+we relocate objects?)
+
+##### Example
+
+Let's take `UserGroup` as an example, as it's a very simple message. `UserGroup`'s interface as it is
+currently defined in `api/types`:
+
+```go
+// UserGroup specifies an externally sourced group.
+type UserGroup interface {
+	ResourceWithLabels
+
+	// GetApplications will return a list of application IDs associated with the user group.
+	GetApplications() []string
+	// SetApplications will set the list of application IDs associated with the user group.
+	SetApplications([]string)
+}
+```
+
+For this message, I propose we create an internal implementation for the above interface, along
+with a builder to completely abstract the internal representation of the user group object:
+
+```go
+type UserGroupBuilder struct {
+  applications []string
+}
+
+func NewUserGroupBuilder() *UserGroupBuilder {
+  return &UserGroupBuilder{}
+}
+
+func (u *UserGroupBuilder) Applications(applications []string) *UserGroupBuilder {
+  u.applications = applications
+  return u
+}
+
+func (u *UserGroupBuilder) Build() (UserGroup, error) {
+  ug := &userGroupImpl{
+    spec: &userGroupSpec{
+      applications: u.applications
+    }
+  }
+  
+  if err := ug.CheckAndSetDefaults(); err != nil {
+    return nil, trace.Wrap(err)
+  }
+
+  return ug, nil
+}
+
+type userGroupSpec struct {
+  Applications []string `json:"applications" yaml:"applications"`
+}
+
+type userGroupImpl struct {
+  // ResourceHeader will contain the version, kind, and metadata fields.
+  ResourceHeader
+
+  Spec *userGroupSpec `json:"spec" yaml:"spec"`
+}
+
+func (u *userGroupImpl) CheckAndSetDefaults() error {
+  ...
+}
+
+func (u *userGroupImpl) GetApplications() string {
+  return u.Spec.Applications
+}
+
+func (u *userGroupImpl) SetApplications(applications []string) {
+  u.Spec.Applications = applications
+}
+...
+```
+
+To go between the protobuf message and the internal implementation, we should have:
+
+```go
+func FromV1(msg proto.UserGroupV1) (types.UserGroup, error) {
```

Comment:
Assuming `FromV1` is in an appropriately-named package (like `usergroup`) then it's a good name. If we are to shove all conversions in a single package (like `api/types`), then we'll need much less-ambiguous names.

Reply:
Let's also avoid putting everything in a single protobuf package, this time 😁

```suggestion
func FromV1(msg usergroupv1.UserGroupV1) (types.UserGroup, error) {
```

Reply:
This is a good chance to take advantage of the non-legacy structure and split the protos a bit more. Huge +1.